### PR TITLE
Fix validation dataloader worker retention

### LINKED ISSUE #25827: Validation retains InfiniteDataLoader workers and prefetched batches between epochs
## GPU comparison — 2026-09-06

[Fresh measurements on PR #26014](https://github.com/ultralytics/ultralytics/pull/26014#issuecomment-5555193776) confirm that the current finite/persistent design drains validation queues. The 1920px small-data check retained about 64 MiB less shared memory with identical predictions. Matched 20-epoch training took 4.8% longer on COCO8 and 11.7% longer on COCO128, with identical per-epoch metrics. Recommend no default loader change from these results alone; the original GB10 memory-pressure workload remains unverified.

## Current status — 2026-09-05

- [x] Lower validation prefetch shipped in #26057 (8.4.139): [current `build_dataloader`](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/build.py) uses **2 for unshuffled loaders**, retaining 4 for shuffled training. The original report below describes 8.4.120 and prefetch 4; its 80-batch/~27-GiB queue estimate is historical (at depth 2, the same raw-image estimate is 40 batches/~13.2 GiB, before overhead).
- [ ] The reported GB10/1920px/batch16 memory-pressure case is **not yet reproduced on current main**. This session's real-image Linux/macOS worker check confirms workers remain alive after a pass and terminate at `close()`, with prefetch 2. It does not reproduce the original 95–100-GB pressure.
- [ ] #26014 remains open. Its current proposal is finite validation iteration with **persistent workers**, not unconditional worker teardown. Validate its current head against the current prefetch-2 baseline on the reported GB10 workload, measuring peak/settled PSS, epoch wall time, and unchanged metrics together.

Preserve the explicit maintainer decision to reuse validation workers during training. Historical matched measurements of the earlier #26014 teardown version showed unconditional teardown/recreation increased wall time ~25.3% and reduced throughput ~20.2%; **recommend no action on that rejected approach**. Later #26014 historical GPU trials were workload-dependent (COCO8 ~5.7% faster, COCO128 ~3.6% slower), not a universal speedup. Reducing `workers` and/or image/batch size is a current memory-pressure workaround. The remaining issue is bounded resource use under the reported configuration, not simply that a persistent worker is alive.

---

### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Val

### General Summary

During detection training, validation uses an `InfiniteDataLoader`. Its worker processes and prefetched batches remain alive after a validation pass completes.

Detection validation also requests twice the configured worker count. With large images, this can retain tens of gigabytes of host memory, saturate the CPU with memory reclaim/compaction, and leave the GPU idle between epochs. There is no exception or traceback. The failure mode is excessive resource retention.

### Actual Behavior

For the following training configuration:

* workers=15
* batch=16
* imgsz=1920
* cache=False
* 20 host CPUs
* NVIDIA GB10 with unified memory
    I observed:
* 15 training pt_data_worker processes
* 20 validation workers, because validation requests workers * 2, capped by the host CPU count
* Approximately 95–100 GB total memory usage
* 100% CPU utilization after validation completed
* kcompactd0 consuming a full CPU core
* GPU mostly idle during the resulting memory pressure
* Validation worker processes still alive after validation exited

Memory Breakdown

The training parent process had approximately:

* 38.6 GiB PSS
* Approximately 36.8 GiB shared-memory PSS
* Approximately 29.8 GiB NVIDIA allocation
    Representative training workers consumed approximately 1.7 GiB PSS each.
    This occurred with cache=False, so the memory usage was not caused by dataset caching.

### Validation Prexetch Memory

At a validation batch size of 32, a raw uint8 image batch at 1920×1920 is approximately 337.5 MiB per batch.
With 20 workers and a prefetch factor of 4, validation may queue as many as 20 workers × 4 batches/worker = 80 batches.

This corresponds to approximately 80 × 337.5 MiB ≈ 27 GiB of raw image data alone, before accounting for multiprocessing, shared-memory, tensor, and Python overhead.
The combination of persistent validation workers and aggressive prefetching therefore creates substantial retained memory pressure even after the validation pass has completed.

### Expected Behavior

* Training may continue using InfiniteDataLoader and persistent workers.
* Validation should use a normal finite `torch.utils.data.DataLoader`.
* Validation workers and prefetch queues should terminate after validation.
* workers=N should not unexpectedly create 2N validation workers.
* Validation should use a conservative prefetch factor to avoid large queued batches when training with high-resolution images.

### Environment

 ```text
  Ultralytics 8.4.120
  Python 3.12.13
  PyTorch 2.13.0+cu130
  Torchvision 0.28.0
  CUDA 13.0

  OS          Linux 6.17.0-1029-nvidia, aarch64
  CPU count   20
  RAM         121.7 GB
  GPU         NVIDIA GB10, 124610 MiB
  GPU count   1

  numpy                 1.26.4
  opencv-python         4.11.0.86
  pillow                12.3.0
  psutil                7.2.2
  polars                1.43.2
  ultralytics-thop      2.1.6
  ultralytics-platform  0.1.5

### Minimal Reproducible Example

  ```python
  from ultralytics.data.build import build_dataloader

  loader = build_dataloader(
      range(64),
      batch=8,
      workers=2,
      shuffle=False,
      device="cpu",
  )

  workers = list(loader.iterator._workers)

  try:
      print("loader type:", type(loader).__name__)
      list(loader)

      alive = [worker.is_alive() for worker in workers]
      print("workers alive after full iteration:", alive)
      print("worker PIDs:", [worker.pid for worker in workers])

      assert not any(alive), "Workers remain alive after the finite pass"
  finally:
      loader.close()
```

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by danielmorozoff**
### General
Closes #25827.

Validation currently uses InfiniteDataLoader, retaining worker processes and prefetched batches after each validation pass. Detection validation also doubles the configured worker count, which can create severe host or unified-memory pressure with high-resolution images.

This PR preserves worker reuse for training while moving validation to standard finite PyTorch dataloaders.

## Changes

* Add an infinite option to build_dataloader(), defaulting to True to preserve existing behavior.
* Use finite dataloaders for detection, classification, and YOLOE validation.
* Stop doubling the configured detection validation worker count.
* Keep the larger prefetch queue for infinite training loaders while allowing finite validation loaders to use PyTorch’s default prefetch depth.
* Add regression coverage for loader selection, prefetch depth, and validation worker configuration.

## Tests

Added regression tests covering:

* Finite versus infinite dataloader selection and prefetch behavior.
* Detection validation respecting the configured worker count.
* Detection validation explicitly disabling worker reuse.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fix validation dataloader worker retention by using finite PyTorch dataloaders during validation while preserving worker reuse for training.

### 📊 Key Changes
- Added an `infinite` option to `build_dataloader` to select between reusable infinite loaders and standard finite `DataLoader` instances.
- Updated detection, classification, and YOLOE validation paths to disable worker reuse.
- Removed validation worker-count multiplication in detection training.
- Added tests covering finite-loader prefetch behavior and validation worker configuration.

### 🎯 Purpose & Impact
- Prevents validation workers from being retained unnecessarily after finite validation runs.
- Reduces worker-related overhead and potential resource retention during validation.
- Preserves the larger prefetch queue and persistent worker behavior for training performance.
- Improves consistency across detection, classification, and YOLOE validation workflows.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @danielmorozoff, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. This is an automated message, and an engineer will assist with your PR if needed. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The loader selection preserves infinite worker reuse for training, switches all affected validation paths to finite PyTorch DataLoaders, retains distributed sampler behavior, and removes the detection validation worker multiplier. The added tests cover the key regression points.

**Review by Y-T-G — CHANGES_REQUESTED** [review]
`prefetch_factor` and worker count changes wouldn't be accepted as they are intended behavior, not bugs.

The only real bug here is the cleanup issue for non training validation. You just to need add an explicit `close()` call on the dataloader to fix that, scoped to non-training validation.

**Comment by danielmorozoff** [comment]
> `prefetch_factor` and worker count changes wouldn't be accepted as they are intended behavior, not bugs.
> 
> The only real bug here is the cleanup issue for non training validation. You just to need add an explicit `close()` call on the dataloader to fix that, scoped to non-training validation.

Understood on preserving the existing worker count and `prefetch_factor`.

To confirm scope: this issue concerns validation invoked during training. In that path, `BaseValidator.self.training` is `True`, so cleanup scoped to not `self.training` leaves validation workers and prefetched batches alive between epochs. Should the loader instead be closed after every validation pass and recreated before the next pass, while preserving the existing worker count and `prefetch_factor`?

**Comment by Y-T-G** [comment]
> To confirm scope: this issue concerns validation invoked during training.

The workers are intended to remain alive between epochs. The PR itself was originally scoped to validation outside training:

> This PR preserves worker reuse for training while moving validation to standard finite PyTorch dataloaders.

I am not sure why you're now saying it concerns training instead

**Comment by danielmorozoff** [comment]
By “preserves worker reuse for training,” I meant the training `dataloader (mode="train")`, not the validation dataloader when invoked by `BaseTrainer`.  

The issue is concerned with validation during training: the report describes validation workers and prefetched batches remaining alive after an epoch’s validation pass. This leads to memory pressure and for larger models/images - you see memory explode.

Its expected behavior distinguishes the two loaders:  
* The training loader may remain persistent. 
* The validation loader should release its workers after each pass.  

Therefore, cleanup guarded by `not self.training` would address standalone validation only, not the behavior reported in #25827.  

If retaining validation workers between training epochs is also intended behavior, I can re-scope this PR to standalone validation cleanup. However it should not resolve the reported memory pressure  in  #25827.

**Comment by glenn-jocher** [comment]
Thanks for clarifying. `self.test_loader` is intentionally reused across training epochs, so please do not close or recreate it after each validation pass. Re-scope this PR to standalone validation cleanup while preserving the existing worker count and `prefetch_factor`; #25827 should remain open for further profiling of the reported memory growth.

**Review by glenn-jocher — CHANGES_REQUESTED** [review]
This is a useful dataloader-lifecycle fix, but the added test_detection_validation_dataloader_limits_workers function is a mock/monkeypatch test rather than coverage of the real worker lifecycle. Please delete that test function and any imports used only by it. The contributor fork has maintainer edits disabled, so I cannot make that deletion directly.

**Comment by danielmorozoff** [comment]
@glenn-jocher I will remove the test- so do you guys want me to remove the lifecycle fix or not?

**Comment by glenn-jocher** [comment]
Thanks for removing the mock-based test. After rechecking the revised head against the actual loader lifecycle, the remaining change still replaces validation-during-training loaders with finite loaders, reduces the established validation worker count, and lowers their prefetch depth. Those are intentional reuse/performance behaviors: the trainer retains its validation loader between epochs and closes it during teardown, while standalone Model.val() owns only a local validator whose InfiniteDataLoader already closes when released. This therefore changes intended behavior without isolating a cleanup defect. I'm closing this rather than adding a parallel loader-selection mode; a future change should start from a reproducible leak outside those existing ownership paths.
